engine.py

```diff
@@ -197,6 +211,153 @@ def log_error(self, msg):
         
     
     ##########################################################################################
+    # panel interfaces
+    
+    def _generate_panel_id(self, dialog_name, bundle):
+        """
+        Given a dialog name and a bundle, generate a Nuke panel id.
+        This panel id is used by Nuke to identify and persist the panel.
+        
+        This will return something like 'shotgun_tk_multi_loader2_main'
+        
+        :param dialog_name: An identifier string to identify the dialog to be hosted by the panel
+        :param bundle: The bundle (e.g. app) object to be associated with the panel
+        :returns: a unique identifier string 
+        """
+        panel_id = "shotgun_%s_%s" % (bundle.name, dialog_name)
+        # replace any non-alphanumeric chars with underscores
+        panel_id = re.sub("\W", "_", panel_id)
+        panel_id = panel_id.lower()
+        self.log_debug("Unique panel id for %s %s -> %s" % (bundle, dialog_name, panel_id))
+        return panel_id 
+        
+    def _panel_factory_callback(self, panel_id):
+        """
+        Given a panel id, generate a panel UI.
+        
+        This method is intended to be executed by nuke itself as a callback
+        that runs when panels are being restored, either at startup or 
+        at UI profile switch.
+        
+        :param panel_id: Unique identifier string for the panel
+        :returns: panel widget object 
+        """
+        # note: Nuke silently consumes any exceptions within the callback system
+        # so add a catch all exception handler around this to make sure
+        # we properly log any output 
+        panel_widget = None
+        try:
+            # note! not using the import as this confuses nuke's calback system 
+            import tk_nuke
+        
+            self.log_debug("Panel Factory Callback: Generating UI for panel id '%s'." % panel_id)
+            
+            # the callback is registered at the same time as the panel callback,
+            # so we can assume that our _panels lookup data structure has a matching
+            # object present
+            panel = self._panels[panel_id]
+            # create a new panel widget
+            panel_widget = tk_nuke.NukePanelWidget(panel["title"], panel_id, panel["widget_class"])
+            # and add it to the current pane context (nuke handles this state)
+            panel_widget.addToPane()
+            
+        except Exception, e:
+            self.log_exception("Could not generate panel UI for panel id '%s'" % panel_id)
```

Comment:
We [are](https://github.com/shotgunsoftware/tk-maya/pull/7/files#r34182112) checking for engine not being ```None```in the maya engine but not here. Is this a bug? Another argument in favor of refactoring this logic in a base class.